meminfo.c

```diff
@@ -57,7 +57,97 @@ static long long available_guesstimate(const char* buf)
+    // As of writing there is no spec for it available to us, so we do not
+    // know whether the columns have guaranteed fixed width.
+    //
+    // So we scan for the field name (e.g. "size") line and parse the second
+    // number.
+
+    long long result;
+
+    char* hit = strstr(buf, search_term);
+    if (hit == NULL) {
+        warn("parse_meminfo: arcstats does not contain size field\n");
+        return -1;
+    }
+
+    // ` ` skips spaces, `%*u` ignores the `type` field.
+    int matches = sscanf(hit + strlen(search_term), " %*u %lld", &result);
+    if (matches < 1) {
+        warn("parse_meminfo: unexpected /proc/spl/kstat/zfs/arcstats contents in size line\n");
+        return -1;
+    }
+
+    return result;
+}
+
+/* If ZFS is in use, compute the part of its cache that is memory-reclaimable.
+ * It currently does not count to `MemAvailable`, like the Linux buffer cache
+ * of other file systems does, see:
+ *   * https://github.com/openzfs/zfs/issues/10255
+ * Returns the number of Bytes reclaimable, or 0 if ZFS is not in use.
+ * This function either returns valid data or kills the process
+ * with a fatal error. */
+long long parse_zfs_arcstats_bytes()
+{
+    long long zfs_available_bytes = 0;
+
+    // In contrast to `fd` for `/proc/meminfo` (a file that should always exist)
+    // this `fd` is not `static` because zfs is a kernel module that can be
+    // loaded/unloaded at runtime, so `arcstats` may not always exist.
+    FILE* fd = NULL;
+    fd = fopen("/proc/spl/kstat/zfs/arcstats", "r");
+    if (fd == NULL && errno != ENOENT) {
+        fatal(106, "could not open /proc/spl/kstat/zfs/arcstats: %s\n", strerror(errno));
+    }
+    if (fd != NULL) {
+        // On Linux 4.19 with ZoL 0.8.3, "wc -c /proc/spl/kstat/zfs/arcstats"
+        // counts 4354 bytes.
+        // 8192 should be enough for the foreseeable future.
+        char arcstats_buf[8192] = { 0 };
+
+        size_t len = fread(arcstats_buf, 1, sizeof(arcstats_buf) - 1, fd);
+        if (ferror(fd)) {
+            fatal(107, "could not read /proc/spl/kstat/zfs/arcstats: %s\n", strerror(errno));
+        }
+        if (len == 0) {
+            fatal(108, "could not read /proc/spl/kstat/zfs/arcstats: 0 bytes returned\n");
+        }
+
+        long long arcstats_size_bytes = get_arcstats_entry("\nsize ", arcstats_buf);
+        if (arcstats_size_bytes > 0) {
+            zfs_available_bytes += arcstats_size_bytes;
+            // The amount of data specified by the following fields is not
+            // reclaimable, so we deduct them:
+            // * `c_min`: https://github.com/openzfs/zfs/wiki/ZFS-on-Linux-Module-Parameters#zfs_arc_min
+            // * `arc_meta_min`: https://github.com/openzfs/zfs/wiki/ZFS-on-Linux-Module-Parameters#zfs_arc_meta_min
+            long long arcstats_c_min_bytes = get_arcstats_entry("\nc_min ", arcstats_buf);
+            if (arcstats_c_min_bytes > 0) {
+                zfs_available_bytes -= arcstats_c_min_bytes;
```

Comment:
This value can become negative! You have to check

    zfs_available_bytes > arcstats_c_min_bytes

first. See https://gist.github.com/rfjakob/0a674048b9efd4b985a70107d64f1110 for example when it would get negative